schema/impl.go

```diff
@@ -180,3 +196,123 @@ func (rc *realCluster) addPod(pod_name string, pod_uid string, namespace *Namesp
 
 	return pod_ptr
 }
+
+// updateInfoType updates the metrics of an InfoType from a ContainerElement.
+// updateInfoType returns the latest timestamp in the resulting TimeStore
+// updateInfoType does not fail if a single ContainerMetricElement cannot be parsed
+func (rc *realCluster) updateInfoType(info *InfoType, ce *cache.ContainerElement) (time.Time, error) {
+	var latest_time time.Time
+
+	if ce == nil {
+		return latest_time, fmt.Errorf("cannot update InfoType from nil ContainerElement")
+	}
+	if info == nil {
+		return latest_time, fmt.Errorf("cannot update a nil InfoType")
+	}
+
+	for _, cme := range ce.Metrics {
+		stamp, err := rc.parseMetric(cme, info.Metrics)
+		if err != nil {
+			glog.V(2).Infof("failed to parse ContainerMetricElement: %s", err)
+			continue
+		}
+		latest_time = latestTimestamp(latest_time, stamp)
+	}
+	return latest_time, nil
+}
+
+// addMetricToMap adds a new metric (time-value pair) to a map of TimeStores.
+// addMetricToMap accepts as arguments the metric name, timestamp, value and the TimeStore map
+func (rc *realCluster) addMetricToMap(metric string, timestamp time.Time, value uint64, dict map[string]*store.TimeStore) error {
+	point := store.TimePoint{
+		Timestamp: timestamp,
+		Value:     value,
+	}
+	if val, ok := dict[metric]; ok {
+		ts := *val
+		err := ts.Put(point)
+		if err != nil {
+			return fmt.Errorf("failed to add metric to TimeStore: %s", err)
+		}
+	} else {
+		new_ts := rc.tsConstructor()
+		err := new_ts.Put(point)
+		if err != nil {
+			return fmt.Errorf("failed to add metric to TimeStore: %s", err)
+		}
+		dict[metric] = &new_ts
+	}
+	return nil
+}
+
+// parseMetric populates a map[string]*TimeStore from a ContainerMetricElement
+// parseMetric returns the ContainerMetricElement timestamp, iff successful.
+func (rc *realCluster) parseMetric(cme *cache.ContainerMetricElement, dict map[string]*store.TimeStore) (time.Time, error) {
+	zeroTime := time.Time{}
+	if cme == nil {
+		return zeroTime, fmt.Errorf("cannot parse nil ContainerMetricElement")
+	}
+	if dict == nil {
+		return zeroTime, fmt.Errorf("cannot populate nil map")
+	}
+
+	timestamp := cme.Stats.Timestamp
+	if cme.Spec.HasCpu {
+		// Add CPU Limit metric
+		cpu_limit := cme.Spec.Cpu.Limit
+		err := rc.addMetricToMap("cpu/limit", timestamp, cpu_limit, dict)
```

Comment:
Better to have sub-structure, which will help with derived stats. Store spec and stats separately in a composite struct.